chain/network/src/tests/cache_edges.rs

```diff
@@ -1,311 +1,221 @@
-        assert_eq!(
-            active_edges,
-            self.routing_table.raw_graph.compute_total_active_edges() as usize
-        );
-        assert_eq!(on_memory.len(), self.routing_table.edges_info.len());
-
-        // Check for peers on disk
-        let mut total_peers = 0;
-        for (peer, nonce) in self.store.iter(DBCol::PeerComponent) {
-            total_peers += 1;
-
-            let peer = PeerId::try_from_slice(peer.as_ref()).unwrap();
-            let nonce = u64::try_from_slice(nonce.as_ref()).unwrap();
-            let peer_ix = self.rev_peers.get(&peer).unwrap();
-            let res = on_disk_peers.get(peer_ix).unwrap();
-            assert_eq!(*res, nonce);
-        }
-        assert_eq!(total_peers, on_disk_peers.len());
-
-        // Check for edges on disk
-        let mut total_nonces = 0;
-        for (nonce, edges) in self.store.iter(DBCol::ComponentEdges) {
-            total_nonces += 1;
-
-            let nonce = u64::try_from_slice(nonce.as_ref()).unwrap();
-
-            let edges = Vec::<Edge>::try_from_slice(edges.as_ref()).unwrap();
-            let current_edges = on_disk_edges.get(&nonce).unwrap();
-
-            assert_eq!(edges.len(), current_edges.len());
+    fn new() -> Self {
+        let mut rng = util::make_rng(87927345);
+        let clock = time::FakeClock::default();
+        let me = data::make_peer_id(&mut rng);
+        let store = create_test_store();
 
-            for edge in edges.iter() {
-                let edge_description = self.get_edge_description(edge);
-                assert!(current_edges.contains(&edge_description));
-            }
-        }
-        assert_eq!(total_nonces, on_disk_edges.len());
+        let graph = Arc::new(RwLock::new(routing::GraphWithCache::new(me.clone())));
+        Self { rng, clock, graph, store, _system: actix::System::new() }
     }
 
-    fn add_edge(&mut self, peer0: usize, peer1: usize, nonce: u64) {
-        let peer0 = self.get_peer(peer0).clone();
-        let peer1 = self.get_peer(peer1).clone();
-        let edge = Edge::new(peer0, peer1, nonce, Signature::default(), Signature::default());
-        self.routing_table.add_verified_edges_to_routing_table(vec![edge]);
+    fn component_store(&self) -> routing::ComponentStore {
+        routing::ComponentStore::new(self.store.clone())
     }
 
-    fn update_routing_table(&mut self) {
-        self.routing_table.recalculate_routing_table();
-        self.routing_table.prune_edges(Prune::OncePerHour, DELETE_PEERS_AFTER_TIME);
+    fn new_actor(&self) -> routing::actor::Actor {
+        routing::actor::Actor::new(self.clock.clock(), self.store.clone(), self.graph.clone())
     }
-}
 
-#[test]
-fn empty() {
-    let _system = System::new();
-
-    let mut test = RoutingTableTest::new();
-    test.check(vec![], vec![], vec![]);
-    assert_eq!(test.routing_table.next_available_component_nonce, 0);
+    fn check(&mut self, want_mem: Vec<Edge>, want_db: Vec<routing::Component>) {
```

Comment:
why do you consider into() ugly? In this case it is roughly the same, but in general I think I would like sometimes helper method to expose more advanced types for arguments to indicate how they are treated internally. And using into() to specify those arguments values consistently, no matter what the underlying type, sounds like a way to go to me.

Reply:
Two bits:

1, why I don't like `.into()`

The specific problem with `.into()` I see is that, at the call-site, it's hard to tell what's going one. The thing is converted into *something*, but to figure that out you need to run type inference in your head. That's why I usually prefer typing `X::from(y)` rather that `y.into()` -- more words, but less work for the reader. The more general rule is that Rust type inference is a bit too powerful, and it's more readable to give a bit more type info than strictly necessary. 

2, why &[] over Vec<>

The general rule of thumb is not try *abstract* over the ownership, but to expose it in the API. If you consume the thing, take `T`, even if most callers have `&T` and would have to `.clone()`. 

In this case, we want to compare with wants, so, logically, we don't need to mutate wants. We kinda do mutate them internally, but that's because we are too lazy to implement the comparison API based only on references, not because the operation fundamentally requires ownership.

Then, this is the test, and for tests, contrary to what I was just saying ( :) ), it sometimes makes sense to save some characters at the call-site. Passing-in `&[]` is easier than passing a `Vec` or a `HashSet`. 

So, kinda both "we only *need* references" and "we want easy call-site" point towards `&[T]` for me, and outweigh "well, actually we need `T` intenrally to make `Eq` work".

Reply:
>I actually need to copy/consume the input to normalize it for comparisons. And &[] would be a tradeoff to make the call site more concise, but vec![] would be fine as well. Is that correct?

Not exactly. There are two ownership semantics that API have:

* an ideal ownership  which would you have with a perfect implementation
* a de-facto ownership semantics you get with the current implementation

In this case, the ideal ownership is `&T`, the de-facto ownership is `T`. My rule of thumb is to follow the ideal ownership.

Reply:
ok, I assumed &[] to be the ideal ownership and removed all cloning of edges within check() function (for Components I'd have to define another type to represent the normalized ones storing references, or tighten access to the current type to allow only for normalized values). Is that better?

Reply:
Yeah!